stbt/keyboard.py

```diff
@@ -0,0 +1,168 @@
+        target of ``navigate_to``).
+
+    :type mask: str or `numpy.ndarray`
+    :param str mask:
+        A mask to use when calling `stbt.press_and_wait` to determine when the
+        current selection has finished moving. If the search page has a
+        blinking cursor you need to mask out the region where the cursor can
+        appear, as well as any other regions with dynamic content (such as a
+        picture-in-picture with live TV). See `stbt.press_and_wait` for more
+        details about the mask.
+
+    :type navigate_timeout: int or float
+    :param navigate_timeout: Timeout (in seconds) for ``navigate_to``. In
+        practice ``navigate_to`` should only time out if you have a bug in your
+        ``graph`` state machine specification.
+
+    """
+
+    def __init__(self, page, graph, mask=None, navigate_timeout=20):
+        self.page = page
+        self.G = nx.parse_edgelist(graph.split("\n"),
+                                   create_using=nx.DiGraph(),
+                                   data=[("key", str)])
+        nx.relabel_nodes(self.G, {"SPACE": " "}, copy=False)
+        _add_weights(self.G)
+
+        self.mask = None
+        if isinstance(mask, numpy.ndarray):
+            self.mask = mask
+        elif mask:
+            self.mask = stbt.load_image(mask)
+
+        self.navigate_timeout = navigate_timeout
+
+    # pylint:disable=fixme
+    # TODO: self.page.selection.text
+    #   This property can return a string, or an object with a ``text``
+    #   attribute that is a string.
+    #
+    # TODO: case sensitive keyboards
+    #   Caps lock can be supported with a graph like this:
+    #       A CAPSLOCK_OFF KEY_LEFT
+    #       CAPSLOCK_OFF CAPSLOCK_ON KEY_OK
+    #       CAPSLOCK_ON a KEY_RIGHT
+    #       a q KEY_UP
+    #       <etc>
+    #   (Other mode changes like ABC -> 123!@# can be supported in the same
+    #   way.)
+    #
+    #   I don't know how best to support SHIFT (temporary mode change):
+    #   - In shifted state KEY_OK will go from "A" to "a". We'd want to use
+    #     press_and_wait before we check the new state. But...
+    #   - In non-shifted state KEY_OK just enters the letter; the screen doesn't
+    #     change otherwise. Currently we don't use press_and_wait here because
+    #     typically the text-box where the text appears is masked out (because
+    #     some UIs have a blinking cursor there) and there is no change anywhere
+    #     else on the screen.
+    #
+    # TODO: Check that KEY_OK adds to the search text? With a property on the
+    #   page object? OCR might be too unreliable for incomplete words. We don't
+    #   use press_and_wait because the text-box might be masked out (some UIs
+    #   have a blinking cursor there).
+
+    def enter_text(self, text):
+        if not text:
+            return  # finished
+        letter = text[0]
+        self.navigate_to(letter)
+        stbt.press("KEY_OK")
+        self.enter_text(text[1:])
```

Comment:
Not doing a full review - this just caught my eye.
Using recursion works here but could lead to a huge stack trace if something went wrong late in the input text. The API already allows this to be written (IMO more simply) using a for loop:
```python
for letter in text:
    self.navigate_to(letter)
    stbt.press("KEY_OK")
```

The other thing would be that you are baking in `KEY_OK`, which won't work for everyone.

Reply:
Good point about the stack traces. Fixed. It's much simpler this way too, thanks.

Regarding the key names: You could make the same argument about KEY_UP/DOWN/LEFT/RIGHT. In the spirit of "convention over configuration" our stance on this is that if you want to use our navigation helpers, use our standard key names.

Thanks for the review.

Reply:
> The other thing would be that you are baking in KEY_OK, which won't work for everyone.

@wmanley suggested that all interaction (keypresses) could go via the page object too. I'm not sure if the page object should implement `press` and `press_and_wait` methods, or something more high-level. And we would fall back to `stbt.press` and `stbt.press_and_wait` if the page object doesn't implement them. But for now I'm not going to do this, until I've seen some cases that would need it.